ddl/column.go

```diff
@@ -337,53 +321,54 @@ func (d *ddl) backfillColumn(t table.Table, columnInfo *model.ColumnInfo, reorgI
 }
 
 func (d *ddl) backfillColumnData(t table.Table, columnInfo *model.ColumnInfo, handles []int64, reorgInfo *reorgInfo) error {
+	defaultVal, _, err := table.GetColDefaultValue(nil, columnInfo)
+	colMap := make(map[int64]*types.FieldType)
+	for _, col := range t.Meta().Columns {
+		colMap[col.ID] = &col.FieldType
+	}
 	for _, handle := range handles {
 		log.Info("[ddl] backfill column...", handle)
-
+		if err != nil {
+			return errors.Trace(err)
+		}
 		err := kv.RunInNewTxn(d.store, true, func(txn kv.Transaction) error {
 			if err := d.isReorgRunnable(txn); err != nil {
 				return errors.Trace(err)
 			}
-
-			// First check if row exists.
-			exist, err := checkRowExist(txn, t, handle)
-			if err != nil {
-				return errors.Trace(err)
-			} else if !exist {
+			tablecodec.EncodeRecordKey(t.RecordPrefix(), handle, 0)
+			rowKey := t.RecordKey(handle, nil)
+			rowVal, err := txn.Get(rowKey)
+			if terror.ErrorEqual(err, kv.ErrNotExist) {
 				// If row doesn't exist, skip it.
 				return nil
 			}
-
-			backfillKey := t.RecordKey(handle, &table.Column{ColumnInfo: *columnInfo})
-			backfillValue, err := txn.Get(backfillKey)
-			if err != nil && !kv.IsErrNotFound(err) {
-				return errors.Trace(err)
-			}
-			if backfillValue != nil {
-				return nil
-			}
-
-			value, _, err := table.GetColDefaultValue(nil, columnInfo)
 			if err != nil {
 				return errors.Trace(err)
 			}
-
-			// must convert to the column field type.
-			v, err := value.ConvertTo(&columnInfo.FieldType)
+			rowColumns, err := tablecodec.DecodeRow(rowVal, colMap)
 			if err != nil {
 				return errors.Trace(err)
 			}
-
-			err = lockRow(txn, t, handle)
+			if _, ok := rowColumns[columnInfo.ID]; ok {
+				// The column is already added by update or insert statement, skip it.
+				return nil
+			}
+			newColumnIDs := make([]int64, 0, len(rowColumns)+1)
+			newRow := make([]types.Datum, 0, len(rowColumns)+1)
+			for colID, val := range rowColumns {
+				newColumnIDs = append(newColumnIDs, colID)
+				newRow = append(newRow, val)
+			}
+			newColumnIDs = append(newColumnIDs, columnInfo.ID)
+			newRow = append(newRow, defaultVal)
+			newRowVal, err := tablecodec.EncodeRow(newRow, newColumnIDs)
```

Comment:
suggestion: change `EncodeRow` to accept a `map[id]Datum`, then we don't have to convert the map into 2 slices.